File: src/wiki/schemas/rules.py
"""Check and lint config models."""

from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

Severity = Literal["error", "warning", "off"]


def coerce_severity(value: object) -> Severity:
    if value is False or value == "false":
        return "off"
    if value is True or value == "true":
        return "error"
    if value in ("error", "warning", "off"):
        return value  # type: ignore[return-value]
    raise ValueError(f"expected error, warning, or off, got {value!r}")


class CheckConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    missing_layout_file: Annotated[Severity, Field(default="error")] = "error"
    frontmatter_schema: Annotated[Severity, Field(default="error")] = "error"
    missing_schema_ref: Annotated[Severity, Field(default="error")] = "error"
    remote_schema_refs: Literal["allow", "deny", "allowlist"] = "allow"
    remote_schema_hosts: list[str] = Field(default_factory=list)

    @field_validator("missing_layout_file", "frontmatter_schema", "missing_schema_ref", mode="before")
    @classmethod
    def _validate_severity(cls, value: object) -> Severity:
        return coerce_severity(value)

    @field_validator("remote_schema_refs", mode="before")
    @classmethod
    def _validate_remote_schema_refs(cls, value: object) -> str:
        if value is None:
            return "allow"
        if not isinstance(value, str):
            raise ValueError(f"expected allow, deny, or allowlist, got {value!r}")
        normalized = value.strip().lower()
        if normalized not in {"allow", "deny", "allowlist"}:
            raise ValueError(f"expected allow, deny, or allowlist, got {value!r}")
        return normalized

    @field_validator("remote_schema_hosts", mode="before")
    @classmethod
    def _validate_remote_schema_hosts(cls, value: object) -> list[str]:
        if value is None:
            return []
        if isinstance(value, str):
            text = value.strip()
            return [text] if text else []
        if isinstance(value, list):
            hosts: list[str] = []
            for item in value:
                if not isinstance(item, str):
                    raise ValueError("remote_schema_hosts items must be strings")
                text = item.strip()
                if not text:
                    raise ValueError("remote_schema_hosts items must be non-empty strings")
                hosts.append(text)
            return hosts
        raise ValueError(f"expected remote_schema_hosts string or list, got {value!r}")


class LintConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    broken_links: Annotated[Severity, Field(default="warning")] = "warning"
    filename_pattern: Annotated[Severity, Field(default="warning")] = "warning"
    headings: Annotated[Severity, Field(default="off")] = "off"
    heading_levels: Annotated[Severity, Field(default="off")] = "off"
    duplicate_headings: Annotated[Severity, Field(default="off")] = "off"
    thematic_breaks: Annotated[Severity, Field(default="off")] = "off"
    link_style: Annotated[Severity, Field(default="warning")] = "warning"

    @field_validator(
        "broken_links",
        "filename_pattern",
        "headings",
        "heading_levels",
        "duplicate_headings",
        "thematic_breaks",
        "link_style",
        mode="before",
    )
    @classmethod
    def _validate_severity(cls, value: object) -> Severity:
        return coerce_severity(value)

File: src/wiki/schemas/rules.ts
/**
 * Check and lint rule configuration.
 *
 * Port of `src/wiki/schemas/rules.py`. Both blocks are flat maps of rule name
 * to severity, and the severity coercion is the interesting part: YAML lets a
 * user write `false`, `"false"`, `true`, or `"true"` in a rule slot, and the
 * Python original accepts all four as shorthands for `off` and `error`.
 */

import { ValueError } from "../errors.ts";
import { pyRepr } from "../pyrepr.ts";
import { type ModelSpec, validateModel } from "./model.ts";
import type { ValidationIssue } from "./validation.ts";

/** The three states a rule can be in. */
export type Severity = "error" | "warning" | "off";

/**
 * Coerce a rule value to a severity.
 *
 * Boolean shorthands come first because Python compares with `is False` before
 * the membership test — `1` is *not* accepted as `error`, even though
 * `1 == True`.
 */
export function coerceSeverity(value: unknown): Severity {
  if (value === false || value === "false") return "off";
  if (value === true || value === "true") return "error";
  if (value === "error" || value === "warning" || value === "off") return value;
  throw new ValueError(`expected error, warning, or off, got ${pyRepr(value)}`);
}

/** The `check:` block: document-schema enforcement during build. */
export interface CheckConfig {
  readonly missing_layout_file: Severity;
  readonly frontmatter_schema: Severity;
  readonly missing_schema_ref: Severity;
  readonly remote_schema_refs: "allow" | "deny" | "allowlist";
  readonly remote_schema_hosts: readonly string[];
}

/** The `lint:` block: advisory checks over the wiki corpus. */
export interface LintConfig {
  readonly broken_links: Severity;
  readonly filename_pattern: Severity;
  readonly headings: Severity;
  readonly heading_levels: Severity;
  readonly duplicate_headings: Severity;
  readonly thematic_breaks: Severity;
  readonly link_style: Severity;
}

/** Coerce a `remote_schema_refs` value, defaulting an absent one to `allow`. */
function coerceRemoteSchemaRefs(value: unknown): string {
  if (value === null || value === undefined) return "allow";
  if (typeof value !== "string") {
    throw new ValueError(
      `expected allow, deny, or allowlist, got ${pyRepr(value)}`,
    );
  }
  const normalized = value.trim().toLowerCase();
  if (
    normalized !== "allow" && normalized !== "deny" &&
    normalized !== "allowlist"
  ) {
    throw new ValueError(
      `expected allow, deny, or allowlist, got ${pyRepr(value)}`,
    );
  }
  return normalized;
}

/**
 * Coerce `remote_schema_hosts`.
 *
 * A bare string is one host rather than a per-character list, and a blank entry
 * is rejected instead of silently matching nothing.
 */
function coerceRemoteSchemaHosts(value: unknown): string[] {
  if (value === null || value === undefined) return [];
  if (typeof value === "string") {
    const text = value.trim();
    return text === "" ? [] : [text];
  }
  if (Array.isArray(value)) {
    const hosts: string[] = [];
    for (const item of value) {
      if (typeof item !== "string") {
        throw new ValueError("remote_schema_hosts items must be strings");
      }
      const text = item.trim();
      if (text === "") {
        throw new ValueError(
          "remote_schema_hosts items must be non-empty strings",
        );
      }
      hosts.push(text);
    }
    return hosts;
  }
  throw new ValueError(
    `expected remote_schema_hosts string or list, got ${pyRepr(value)}`,
  );
}

export const checkConfigSpec: ModelSpec = {
  label: "CheckConfig",
  forbidExtra: true,
  fields: [
    ["missing_layout_file", { defaultValue: "error", before: coerceSeverity }],
    ["frontmatter_schema", { defaultValue: "error", before: coerceSeverity }],
    ["missing_schema_ref", { defaultValue: "error", before: coerceSeverity }],
    ["remote_schema_refs", {
      defaultValue: "allow",
      before: coerceRemoteSchemaRefs,
    }],
    ["remote_schema_hosts", {
      factory: () => [],
      before: coerceRemoteSchemaHosts,
    }],
  ],
};

export const lintConfigSpec: ModelSpec = {
  label: "LintConfig",
  forbidExtra: true,
  fields: [
    ["broken_links", { defaultValue: "warning", before: coerceSeverity }],
    ["filename_pattern", { defaultValue: "warning", before: coerceSeverity }],
    ["headings", { defaultValue: "off", before: coerceSeverity }],
    ["heading_levels", { defaultValue: "off", before: coerceSeverity }],
    ["duplicate_headings", { defaultValue: "off", before: coerceSeverity }],
    ["thematic_breaks", { defaultValue: "off", before: coerceSeverity }],
    ["link_style", { defaultValue: "warning", before: coerceSeverity }],
  ],
};

/** The default `check:` block, as `DEFAULT_CHECK_CONFIG`. */
export function defaultCheckConfig(): CheckConfig {
  return coerceCheckConfig({});
}

/** The default `lint:` block, as `DEFAULT_LINT_CONFIG`. */
export function defaultLintConfig(): LintConfig {
  return coerceLintConfig({});
}

/** Validate a `check:` block from raw config input. */
export function coerceCheckConfig(
  raw: unknown,
  loc: readonly (string | number)[] = [],
  issues: ValidationIssue[] = [],
): CheckConfig {
  return validateModel(
    checkConfigSpec,
    raw,
    loc,
    issues,
  ) as unknown as CheckConfig;
}

/** Validate a `lint:` block from raw config input. */
export function coerceLintConfig(
  raw: unknown,
  loc: readonly (string | number)[] = [],
  issues: ValidationIssue[] = [],
): LintConfig {
  return validateModel(
    lintConfigSpec,
    raw,
    loc,
    issues,
  ) as unknown as LintConfig;
}
